### ISSUE TITLE
nx.DiGraph.edge_subgraph creates graph with corrupt pred and succ

### ISSUE THREAD

**Issue by KyleBenson** [networkx/networkx#2370]
The diff below shows the problem in the DiGraph.edge_subgraph function and a potential fix for it: when adding directed edge u,v from the list of edges, it only adds an entry to succ[u] and pred[v], neglecting to add pred[u] and succ[v].  Here's a minimal non-working example (master branch):
```python
g = nx.DiGraph([(0,1),(0,2),(2,3)])
g2 = gg.edge_subgraph([(0,1),(0,2),(2,3)])
print nx.info(g2)
```

This results in:

```python
networkx/classes/digraph.pyc in <genexpr>((n, succs))
    906 
    907         if nbunch is None:
--> 908             nodes_nbrs=( (n,succs,self.pred[n]) for n,succs in self.succ.items())
    909         else:
    910             nodes_nbrs=( (n,self.succ[n],self.pred[n]) for n in self.nbunch_iter(nbunch))

KeyError: 0
```

Here's a quick and dirty fix, though I should note that we probably don't need so many if statements and could add these pred/succ entries at the same time as adding the node entries:

```
diff --git a/networkx/classes/digraph.py b/networkx/classes/digraph.py
index 9e08653..6470a09 100644
--- a/networkx/classes/digraph.py
+++ b/networkx/classes/digraph.py
@@ -1334,6 +1334,10 @@ class DiGraph(Graph):
                 H.succ[u] = H.adjlist_inner_dict_factory()
             if v not in H.pred:
                 H.pred[v] = H.adjlist_inner_dict_factory()
+            if u not in H.pred:
+                H.pred[u] = H.adjlist_inner_dict_factory()
+            if v not in H.succ:
+                H.succ[v] = H.adjlist_inner_dict_factory()
             # Copy the edge attributes.
             H.edge[u][v] = self.edge[u][v]
             H.pred[v][u] = self.pred[v][u]
```

**Comment by dschult** [comment]
Perhaps I am confused, but for a DiGraph adding the edge (u,v) should only add to succ[u] and pred[v]. If we add both directions then we have a Graph, not a DiGraph.

**Comment by KyleBenson** [comment]
Thanks for the reply and sorry for the confusion.  Please allow me to clarify.  We are not adding the edge in both directions (edge[u][v] and edge[v][u]) like in a Graph.  Rather we are really adding the nodes u and/or v for the first time and so must allocate space in the pred/succ dicts.  Indeed these pred and succ statements could be made a few lines above when the nodes are explicitly added for the first time.

The bug arises when an algorithm (such as nx.info or DFS/BFS) sees a node and inspects if it has any successors and/or predecessors (see above where we get n,succs from the dict.items() and then try to use the key n to do pred[n]).  Perhaps it would be better to use try/except statements in such algorithms (or public APIs rather than private members), but I just wanted to propose a quick simple fix for the sake of completeness :)

**Comment by dschult** [comment]
Yes-- that's a bug all right....  Adding a node and only adding half of the pred/succ structure is a problem.   Thanks for identifying and reporting it!

**Comment by jfinkels** [comment]
Fixed in pull request #2373. Thanks!
